pkg/metaclient/kvclient/etcdkv/etcd_client.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	return etcdOps
+}
+
+func (c *etcdImpl) getEtcdOp(op metaclient.Op) clientv3.Op {
+	opts := c.getEtcdOptions(op)
+	switch {
+	case op.IsGet():
+		return clientv3.OpGet(string(op.KeyBytes()), opts...)
+	case op.IsPut():
+		return clientv3.OpPut(string(op.KeyBytes()), string(op.ValueBytes()), opts...)
+	case op.IsDelete():
+		return clientv3.OpDelete(string(op.KeyBytes()), opts...)
+	case op.IsTxn():
+		ops := op.Txn()
+		etcdOps := make([]clientv3.Op, 0, len(ops))
+		for _, sop := range ops {
+			etcdOps = append(etcdOps, c.getEtcdOp(sop))
+		}
+		return clientv3.OpTxn(nil, etcdOps, nil)
+	}
+
+	panic("unknown op type")
+}
+
+func (c *etcdImpl) Put(ctx context.Context, key, val string) (*metaclient.PutResponse, error) {
+	op := metaclient.OpPut(key, val)
+	etcdResp, err := c.cli.Do(ctx, c.getEtcdOp(op))
+	if err != nil {
+		return nil, cerrors.ErrMetaOpFail.FastGenByArgs(err)
+	}
+
+	putRsp := etcdResp.Put()
+	c.getLatestRev(putRsp.Header)
+	return makePutResp(putRsp), nil
+}
+
+func (c *etcdImpl) Get(ctx context.Context, key string, opts ...metaclient.OpOption) (*metaclient.GetResponse, error) {
+	op := metaclient.OpGet(key, opts...)
+	if err := op.CheckValidOp(); err != nil {
+		return nil, cerrors.ErrMetaOptionInvalid.Wrap(err)
+	}
+
+	etcdResp, err := c.cli.Do(ctx, c.getEtcdOp(op))
+	if err != nil {
+		return nil, cerrors.ErrMetaOpFail.FastGenByArgs(err)
+	}
+
+	getRsp := etcdResp.Get()
+	c.getLatestRev(getRsp.Header)
+	return makeGetResp(getRsp), nil
+}
+
+func (c *etcdImpl) Delete(ctx context.Context, key string, opts ...metaclient.OpOption) (*metaclient.DeleteResponse, error) {
+	op := metaclient.OpDelete(key, opts...)
+	if err := op.CheckValidOp(); err != nil {
+		return nil, cerrors.ErrMetaOptionInvalid.Wrap(err)
+	}
+
+	etcdResp, err := c.cli.Do(ctx, c.getEtcdOp(op))
+	if err != nil {
+		return nil, cerrors.ErrMetaOpFail.FastGenByArgs(err)
+	}
+
+	delRsp := etcdResp.Del()
+	c.getLatestRev(delRsp.Header)
+	return makeDeleteResp(delRsp), nil
+}
+
+func (c *etcdImpl) Do(ctx context.Context, op metaclient.Op) (metaclient.OpResponse, error) {
+	if err := op.CheckValidOp(); err != nil {
+		return metaclient.OpResponse{}, err
+	}
+
+	etcdResp, err := c.cli.Do(ctx, c.getEtcdOp(op))
+	if err != nil {
+		return metaclient.OpResponse{}, cerrors.ErrMetaOpFail.FastGenByArgs(err)
+	}
+
+	switch {
+	case op.IsGet():
+		rsp := etcdResp.Get()
+		c.getLatestRev(rsp.Header)
+		getRsp := makeGetResp(rsp)
+		return getRsp.OpResponse(), nil
+	case op.IsPut():
+		rsp := etcdResp.Put()
+		c.getLatestRev(rsp.Header)
+		putRsp := makePutResp(rsp)
+		return putRsp.OpResponse(), nil
+	case op.IsDelete():
+		rsp := etcdResp.Del()
+		c.getLatestRev(rsp.Header)
+		delRsp := makeDeleteResp(rsp)
+		return delRsp.OpResponse(), nil
+	case op.IsTxn():
+		rsp := etcdResp.Txn()
+		c.getLatestRev(rsp.Header)
+		txnRsp := makeTxnResp(rsp)
+		return txnRsp.OpResponse(), nil
+	default:
+		panic("Unknown op")
+	}
+}
+
+type etcdTxn struct {
+	clientv3.Txn
+	mu  sync.Mutex
```

Comment:
Please make it clear what fields `mu` is intended to protect.

Reply:
Almost all inner members, so I don't specify the fields.